### PR TITLE
Add process group implementation and support set/getpgid, set/getpgrp

### PR DESCRIPTION
**Pull request by jessehui**

### CODE REVIEW

File: src/libos/src/process/do_spawn/spawn_attr.rs [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    __sp: sched_param,
+    __policy: i32,
+}
+
+impl spawnattr_t {
+    pub fn from_raw_ptr(attr_ptr: *const spawnattr_t) -> Result<Self> {
+        check_ptr(attr_ptr)?;
+        let attr = unsafe { *attr_ptr };
+        attr.validate()?;
+        Ok(attr)
```
Comment (guzongmin) [inline src/libos/src/process/do_spawn/spawn_attr.rs:42]: POSIX_SPAWN_SETPGROUP
              Set the process group to the value specified in the spawn-
              pgroup attribute of the object pointed to by attrp.  **If the
              spawn-pgroup attribute has the value 0, the child's process
              group ID is made the same as its process ID.**  If the
              POSIX_SPAWN_SETPGROUP flag is not set, the child inherits the
              parent's process group ID.

Do you need to check if the spawn-pgroup attribute is 0?

File: src/libos/src/process/syscalls.rs [hunk 2]
```diff
@@ -247,11 +255,31 @@ pub fn do_getppid() -> Result<isize> {
...
-pub fn do_getpgid() -> Result<isize> {
-    let pgid = super::do_getpid::do_getpgid();
+pub fn do_getpgid(pid: pid_t) -> Result<isize> {
+    let real_pid = if pid == 0 { do_getpid()? as pid_t } else { pid };
+    let pgid = super::pgrp::do_getpgid(real_pid)?;
     Ok(pgid as isize)
 }
 
+pub fn do_setpgid(pid: i32, pgid: pid_t) -> Result<isize> {
+    if pid < 0 {
```
Comment (jessehui) [inline src/libos/src/process/syscalls.rs:265]: Due to previous comments:
EINVAL: pgid is less than 0 (setpgid(), setpgrp()).
ESRCH: For getpgid(): pid does not match any process. For setpgid(): pid is not the calling process and not a child of the calling process.

File: src/libos/src/process/task/exec.rs [hunk 3]
```diff
@@ -46,6 +46,7 @@ pub fn exec(libos_tid: pid_t, host_tid: pid_t) -> Result<i32> {
         // task may only be modified by this function; so no lock is needed
         do_exec_task(this_thread.task() as *const Task as *mut Task);
     }
+    println!("thread {:?} exit???", this_thread.tid());
```
Comment (guzongmin) [inline src/libos/src/process/task/exec.rs:49]: Can you please replace the print with log?

File: src/libos/src/process/syscalls.rs (hunk 2 above)
Comment (jessehui) [reply src/libos/src/process/syscalls.rs:265]: Done.

File: src/libos/src/process/do_spawn/spawn_attr.rs (hunk 1 above)
Comment (jessehui) [reply src/libos/src/process/do_spawn/spawn_attr.rs:42]: This check is done in `pgrp::do_setpgid` function.

File: src/libos/src/process/task/exec.rs (hunk 3 above)
Comment (jessehui) [reply src/libos/src/process/task/exec.rs:49]: Done.

File: src/libos/src/process/pgrp.rs [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+        }
+        return Ok(0);
+    }
+}
+
+impl Drop for ProcessGrp {
+    fn drop(&mut self) {
+        if self.get_process_number() == 0 {
+            warn!("drop process group");
+            // let pgid = self.pgid();
```
Comment (guzongmin) [inline src/libos/src/process/pgrp.rs:125]: remove those comments

**Review by guzongmin — APPROVED** [review]
LGTM

**Pull request merged**